File: src/content/algorithms/trie/code.py
class TrieNode:  # #@class-def
    def __init__(self):  # #@class-def
        self.children = {}  # #@class-def
        self.is_end = False  # #@mark-end


class Trie:  # #@class-def
    def __init__(self):  # #@init-root
        self.root = TrieNode()  # #@init-root

    def insert(self, word):  # #@insert-op
        node = self.root  # #@traverse
        for char in word:  # #@traverse
            if char not in node.children:  # #@check-exists
                node.children[char] = TrieNode()  # #@create-node
            node = node.children[char]  # #@traverse
        node.is_end = True  # #@mark-end

    def search(self, word):  # #@search-op
        node = self._find_node(word)  # #@search-op
        return node is not None and node.is_end  # #@search-op

    def startsWith(self, prefix):  # #@prefix-op
        return self._find_node(prefix) is not None  # #@prefix-op

    def _find_node(self, s):  # #@traverse
        node = self.root  # #@traverse
        for char in s:  # #@traverse
            if char not in node.children:  # #@check-exists
                return None  # #@check-exists
            node = node.children[char]  # #@traverse
        return node  # #@traverse

    def autocomplete(self, prefix):
        node = self._find_node(prefix)
        if not node:
            return []
        results = []
        self._dfs(node, prefix, results)
        return results

    def _dfs(self, node, current, results):
        if node.is_end:
            results.append(current)
        for char, child in node.children.items():
            self._dfs(child, current + char, results)

File: src/content/algorithms/trie/code.rs
use std::collections::HashMap;

struct TrieNode {  // #@class-def
    children: HashMap<char, TrieNode>,  // #@class-def
    is_end: bool,  // #@mark-end
}

impl TrieNode {
    fn new() -> Self {  // #@class-def
        TrieNode {
            children: HashMap::new(),
            is_end: false,
        }
    }
}

struct Trie {  // #@class-def
    root: TrieNode,  // #@init-root
}

impl Trie {
    fn new() -> Self {  // #@init-root
        Trie {
            root: TrieNode::new(),  // #@init-root
        }
    }

    fn insert(&mut self, word: &str) {  // #@insert-op
        let mut node = &mut self.root;  // #@traverse
        for char in word.chars() {  // #@traverse
            node = node.children  // #@create-node
                .entry(char)
                .or_insert_with(TrieNode::new);  // #@create-node
        }
        node.is_end = true;  // #@mark-end
    }

    fn search(&self, word: &str) -> bool {  // #@search-op
        match self.find_node(word) {  // #@search-op
            Some(node) => node.is_end,  // #@search-op
            None => false,
        }
    }

    fn starts_with(&self, prefix: &str) -> bool {  // #@prefix-op
        self.find_node(prefix).is_some()  // #@prefix-op
    }

    fn find_node(&self, s: &str) -> Option<&TrieNode> {  // #@traverse
        let mut node = &self.root;  // #@traverse
        for char in s.chars() {  // #@traverse
            match node.children.get(&char) {  // #@check-exists
                Some(child) => node = child,  // #@traverse
                None => return None,  // #@check-exists
            }
        }
        Some(node)  // #@traverse
    }

    fn autocomplete(&self, prefix: &str) -> Vec<String> {
        let mut results = Vec::new();
        if let Some(node) = self.find_node(prefix) {
            self.dfs(node, prefix.to_string(), &mut results);
        }
        results
    }

    fn dfs(&self, node: &TrieNode, current: String, results: &mut Vec<String>) {
        if node.is_end {
            results.push(current.clone());
        }
        for (&char, child) in &node.children {
            let mut next = current.clone();
            next.push(char);
            self.dfs(child, next, results);
        }
    }
}
